ocaml/typing/solver.ml

```diff
@@ -34,34 +34,46 @@ let rec find_error (f : 'x -> ('a, 'b) Result.t) : 'x list -> ('a, 'b) Result.t
   | x :: rest -> (
     match f x with Ok () -> find_error f rest | Error _ as e -> e)
 
-module VarSet = Set.Make (Int)
-
 module Solver_mono (C : Lattices_mono) = struct
   type 'a var =
-    { mutable vlower : 'a lmorphvar list
-      (** A list of variables directly under the current variable.
+    { mutable vlower : 'a lmorphvar list;
+          (** A list of variables directly under the current variable.
         Each is a pair [f] [v], and we have [f v <= u] where [u] is the current
         variable.
         TODO: consider using hashset for quicker deduplication *)
-      mutable upper : 'a;
-      (** The precise upper bound of the variable *)
+      mutable upper : 'a;  (** The precise upper bound of the variable *)
       mutable lower : 'a;
-      (** The *conservative* lower bound of the variable.
+          (** The *conservative* lower bound of the variable.
        Why conservative: if a user calls [submode c u] where [c] is
         some constant and [u] some variable, we can modify [u.lower] of course.
        Idealy we should also modify all [v.lower] where [v] is variable above [u].
        However, we only have [vlower] not [vupper]. Therefore, the [lower] of
        higher variables are not updated immediately, hence conservative. Those
        [lower] of higher variables can be made precise later on demand, see
-       [zap_to_floor_try], which is why we decide we don't need [vupper].
-       *)
+       [zap_to_floor_try].
+
+       One might argue for an additional [vupper] field, so that [lower] are
+       always precise. While this might be doable, we note that the "hotspot" of
+       the mode solver is to detect conflict, which is already achieved without
+       precise [lower]. Adding [vupper] and keeping [lower] precise will come
+       at extra cost. *)
+      (* To summarize, INVARIANT:
+         - For any variable [v], we have [v.lower <= v.upper].
+         - Variables that have been fully constrained will have
+         [v.lower = v.upper]. Note that adding a boolean field indicating that
+         won't help much.
+         - For any [v] and [u \in v.vlower], we have [u.upper <= v.upper], but not
```

Comment:
I think this isn't quite right: `u` is a `morphvar`, which doesn't have an `upper`.

Reply:
fixed